contracts/lottery/src/lib.rs

```diff
@@ -0,0 +1,608 @@
+            return Err(Error::MinParticipantsNotSatisfied);
+        }
+
+        let pool = token_client.balance(&env.current_contract_address());
+        let max_range = storage.get::<_, u32>(&DataKey::MaxRange).unwrap();
+        let number_of_elements = storage.get::<_, u32>(&DataKey::NumberOfNumbers).unwrap();
+        let mut thresholds = storage
+            .get::<_, Map<u32, u32>>(&DataKey::Thresholds)
+            .unwrap();
+
+        let drawn_numbers = draw_numbers(&env, max_range, number_of_elements, random_seed);
+        let winners = get_winners(&env, &drawn_numbers, &tickets, &thresholds);
+        let prizes = calculate_prizes(&env, &winners, &mut thresholds, pool);
+        payout_prizes(&env, &token_client, &prizes);
+
+        // store numbers drawn in this lottery
+        let lottery_number = storage.get::<_, u32>(&DataKey::LotteryNumber).unwrap();
+        let mut lottery_results = storage
+            .get::<_, Map<u32, LotteryResult>>(&DataKey::LotteryResults)
+            .unwrap_or(map![&env]);
+
+        lottery_results.set(lottery_number, drawn_numbers);
+        storage.set(&DataKey::LotteryResults, &lottery_results);
+
+        storage.set(&DataKey::LotteryState, &LotteryState::Finished);
+
+        // emit events with won prizes
+        prizes.iter().for_each(|(address, prize)| {
+            let topic = (Symbol::new(&env, "won_prize"), &address);
+            env.events().publish(topic, prize);
+        });
+        Ok(())
+    }
+}
+
+/// Ensures lottery is initialized and not finished.
+/// If not, error is returned.
+///
+/// # Arguments
+///
+/// - `storage` - Contracts data storage
+fn  lottery_must_be_active(storage: &Persistent) -> Result<(), Error> {
+    let lottery_state_opt = storage
+        .get::<_, LotteryState>(&DataKey::LotteryState);
+
+    if lottery_state_opt.is_none() {
+        return Err(Error::NotInitialized);
+    }
+
+    if lottery_state_opt.unwrap() != LotteryState::Active {
+        return Err(Error::NotActive);
+    }
+    Ok(())
+}
+
+/// Calculates the winners of a lottery. There can be several winners
+/// with different prizes according to number of correctly selected numbers and
+/// specified thresholds.
+///
+/// # Returns
+///
+/// A map where number of selected numbers is a key, and
+/// vector of addresses that properly selected this number of numbers is a value
+///
+/// # Arguments
+///
+/// - `env` - The environment for this contract.
+/// - `drawn_numbers` - An array containing numbers selected in this lottery,
+/// - `tickets` - Tickets with players selection
+/// - `thresholds` - Defined thresholds with prizes
+fn get_winners(
+    env: &Env,
+    drawn_numbers: &Vec<u32>,
+    tickets: &Map<Address, Vec<LotteryTicket>>,
+    thresholds: &Map<u32, u32>,
+) -> Map<u32, Vec<Address>> {
+    let mut winners = Map::<u32, Vec<Address>>::new(&env);
+
+    tickets.iter().for_each(|(ticket_address, tickets)| {
+        tickets.iter().for_each(|ticket| {
```

Comment:
A loop inside a loop.

Reply:
I've done some testing with different approaches and compared spent fees which gives some sight about the performance. 
The initial approach with nested loops:
```
Player fees:  312307
Admin fees:  1103201
Total fees:  1415508
```
Then I've tried to flatten the data structures and used vectors of pairs instead of maps for storing tickets and winners. That actually decreased the performance as total number of loop iterations increased even if there were none nested loops:
```
Player fees:  394744
Admin fees:  1201939
Total fees:  1596683
```
Finally I've back to the original approach and done slight optimization in searching for winners (added filtering before final inner loop, latest commit) and that gave a slightly better results:
```
Player fees:  312307
Admin fees:  958732
Total fees:  1271039
```
If Soroban vectors and maps would have `FromIter` trait implemented it might be possible to further optimize the code, as then we could use something like
```
  vector.iter().map(...).collect();
```
instead of 
```
  vector.iter().for_each(...);
```